src/v/coproc/script_context_backend.cc

```diff
@@ -189,69 +192,126 @@ process_one_reply(process_batch_reply::data e, output_write_args args) {
+    }
+}
+
+using grouping_t = absl::flat_hash_map<model::ntp, output_write_inputs>;
+
+static grouping_t group_replies(output_write_inputs replies) {
+    grouping_t gt;
+    for (auto& r : replies) {
+        auto [itr, _] = gt.try_emplace(r.source);
+        itr->second.push_back(std::move(r));
+    }
+    return gt;
+}
+
+static ss::future<> process_reply_group(
+  model::ntp source, output_write_inputs reply_group, output_write_args args) {
+    auto found = args.inputs.find(source);
+    if (found == args.inputs.end()) {
+        throw bad_reply_exception(
+          fmt::format("script {} unknown source ntp: {}", args.id, source));
+    }
+    auto src_ptr = found->second;
+    for (auto& e : reply_group) {
+        try {
+            co_await process_one_reply(std::move(e), src_ptr, args);
+        } catch (const bad_reply_exception& ex) {
+            /// A bad reply shouldn't affect the outcome, just ignore
+            vlog(coproclog.error, "Erraneous response detected: {}", ex);
+        }
+    }
+    /// If we've reached here without error, all materialized
+    /// partitions (working under the input) can be promoted, rational is to
+    /// support filter operations, i.e. responses were returned for some
+    /// materialized partitions but not others. If offsets for missing ones
+    /// were not bumped in case of success, there would be an infinite retry
+    /// loop.
+    for (auto& [_, o] : src_ptr->wctx.offsets) {
+        if (o <= src_ptr->rctx.last_read) {
+            /// Omit increasing offets that are ahead of current read, only
+            /// possibility of this case would be a load of a stored offset that
+            /// is far ahead
+            o = src_ptr->rctx.last_read;
+        }
     }
-    auto ofound = ntp_ctx->offsets.find(args.id);
-    vassert(
-      ofound != ntp_ctx->offsets.end(),
-      "Offset not found for script id {} for ntp owning context: {}",
-      args.id,
-      ntp_ctx->ntp());
-    /// Reset the acked offset so that progress can be made
-    ofound->second.last_acked = ofound->second.last_read;
 }
 
 ss::future<>
 write_materialized(output_write_inputs replies, output_write_args args) {
     if (replies.empty()) {
         vlog(
           coproclog.error, "Wasm engine interpreted the request as erraneous");
-    } else {
-        for (auto& e : replies) {
-            co_await process_one_reply(std::move(e), args);
-        }
+        co_return;
+    }
+    grouping_t grs = group_replies(std::move(replies));
+    bool err{false};
+    co_await ss::parallel_for_each(
+      grs, [args, &err](grouping_t::value_type& vt) -> ss::future<> {
+          try {
+              co_await process_reply_group(
+                vt.first, std::move(vt.second), args);
+          } catch (const bad_reply_exception& ex) {
+              vlog(coproclog.error, "No source for reply: {}", ex);
+          } catch (const script_failed_exception& ex) {
+              throw ex;
```

Comment:
`script_failed_exception` is a `coproc::exception` - this handler is redundant.

Reply:
not in this case, because if the handler below it catches this error, then it will fail to percolate up and close the fiber. The idea is to percolate to `pacemaker.cc` where the fiber will be closed and cleaned up